(NFC) Upgrade dependencies to "green commits" for week of 01-23-2023

christopherbate opened the pull request:
Upgrades llvm-project and mlir-hlo project dependencies to the commits from 01-23-2022 per https://github.com/llvm/torch-mlir/issues/1178.

LLVM-Project: 9acc2f37bdfce08ca0c2faec03392db10d1bb7a9 
Mlir-hlo: 4a173356bb1291b97046545429d7851cbc771d88

The major changes in the onnx-mlir codebase consist of:

- All generated accessor methods in operations move to prefixed form (e.g. `op.input()` -> `op.getInput()`).
- Hard-coded uses of `-1` to represent a dynamic dimension in the context of a `ShapedType` (e.g. `TensorType`, `MemRefType`, and derivatives) are removed and replaced with `ShapedType::kDynamic`.
- Replace uses of `mlir::None` with `std::nullopt`
- Replace uses of `llvm::None` with `std::nullopt`.
- Replace uses of `llvm::makeArrayRef` with `llvm::ArrayRef`

Warnings not yet fixed:

  - Warnings due to MLIR `fold` API migration

I labeled this as NFC because it's just trying to upgrade the llvm dependency and there's no functional change. But it affects nearly every file due to getter/setter prefix change in upstream that makes MLIR no longer support the non-prefixed form.

NathanielMcVicar commented on src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp:
```diff
@@ -132,10 +132,13 @@ class ONNXMaxPoolSingleOutOpLoweringToTOSA : public ConversionPattern {
 
-    mlir::ArrayAttr strides = rewriter.getI64ArrayAttr(shapeHelper.strides);
+    auto strides = rewriter.getDenseI64ArrayAttr(shapeHelper.strides);
+    auto kernelShape = rewriter.getDenseI64ArrayAttr(llvm::to_vector(
```
Since kernelShape is already an I64ArrayAttr I believe you can use the helper here:
`auto kernelShape = rewriter.getDenseI64ArrayAttr(extractFromI64ArrayAttr(adaptor.getKernelShapeAttr()));`

AlexandreEichenberger commented:
@jenkins-droid test this please

AlexandreEichenberger commented:
@christopherbate thanks for this massive undertaking. Really appreciated.

Please state that you will follow the DCO, namely that all the code changes are yours and follow the DCO policy. By signing the commit (see README.md toward the bottom), you will be able to pass the DCO automatically. 

Also, we eventually grant privilege to run the Jenkins once you have done a few PRs.

AlexandreEichenberger commented:
Since this PR contains so much changes, I suggest that we merge is as soon as it passes the CIs and there is no glaring errors, as merging other PRs into this one must be quite painful.

We can then take a second look (once its passing and merged) if really all the `-1` were translated into the `ShapedType::kDynamic` by looking at all the uses of both values in their context.

christopherbate commented on src/Builder/ModelInputShaper.cpp:
```diff
@@ -94,7 +94,7 @@ RankedTensorType forceShape(
+    // Make dims unknown (ShapedType::kDynamic) if applicable.
     if (force_dim_dynamic_enabled_ && tensorTy.hasRank()) {
       auto rankedTensorTy = tensorTy.cast<RankedTensorType>();
       auto it = forced_inputs_dims_.find(-1);
```
It looks like it, let me check.

christopherbate replied:
So actually this class is a bridge between the information in the ONNX graph, the environment variable `IMPORTER_FORCE_DYNAMIC` and MLIR Types. 

ONNX uses `-1` and the environment variable also allows use of `-1`, so actually the change to the comment is wrong.

christopherbate commented:
>Since this PR contains so much changes, I suggest that we merge is as soon as it passes the CIs and there is no glaring errors, as merging other PRs into this one must be quite painful.

That would be great, I'll try to get the CI passing asap.

> We can then take a second look (once its passing and merged) if really all the -1 were translated into the ShapedType::kDynamic by looking at all the uses of both values in their context.

I just checked and found more -1 that haven't been changed, but I'm passing all LIT checks. I'll do my best.

christopherbate commented:
Added updates for NNPA dialects. 
Updated LLVM commit hash under `utils/clone-mlir.sh` to fix the CI.

Locally I'm now passing most tests:
```
  Unsupported:   6
  Passed     : 187
  Failed     :   1
```

christopherbate commented:
I've avoided running `clang-format` to make reviewing changes easier. It might be good to hold off on doing that until after merge, then do a separate commit with a bulk clang-format applied.

AlexandreEichenberger commented:
@jenkins-droid test this please

christopherbate commented:
> ERROR: Exception: java.lang.Exception: admin authentication failed
Finished: FAILURE

AlexandreEichenberger commented:
@jenkins-droid test this please

AlexandreEichenberger commented:
For the format, I am not aware how to put down code that fails this CI. My recommendation is to pass it, maybe before merging the code down. The CIs take about 1-2h, so it should't take much additional time.

AlexandreEichenberger reviewed (approved):
Because of the amount of changes, I propose that we put down the code as soon as it passes the CIs. We may want then to look at the changes (esp the one with the -1/kDynamic) to make sure we did not pick a -1 that has nothing to do with the kDynamic, and did not forget a -1 that should use the kDynamic. We don't want to experience massive failure if in the future they changed the value of kDynamic to something else.

AlexandreEichenberger commented:
@christopherbate if you suspect Jenkins failures due to its infrastructure, the person to ping is @gongsu832, who works on Asia's time schedule. Looks that for the moments the CIs are doing ok. Note that you can run all of the CIs also locally on your machine. `onnx-check-lit`, `onnx-check-backend`, `onnx-check-backend-dynamic` and `onnx-check-numerical` are usually a good coverage of the total CIs.

AlexandreEichenberger commented:
Unfortunately an issue with dynamic shapes (from linux x86)
```
Test case generation with auto pad = VALID or SAME and dynamic.

- convolution implementation correctness
error: invalid tensor dimension size
TestConv: /workdir/llvm-project/mlir/include/mlir/IR/StorageUniquerSupport.h:152: static ConcreteT mlir::detail::StorageUserBase<ConcreteT, BaseT, StorageT, UniquerT, Traits>::get(mlir::MLIRContext*, Args ...) [with Args = {llvm::ArrayRef<long int>, mlir::Type, mlir::Attribute}; ConcreteT = mlir::RankedTensorType; BaseT = mlir::TensorType; StorageT = mlir::detail::RankedTensorTypeStorage; UniquerT = mlir::detail::TypeUniquer; Traits = {mlir::SubElementTypeInterface::Trait, mlir::ShapedType::Trait}]: Assertion `succeeded(ConcreteT::verify(getDefaultDiagnosticEmitFn(ctx), args...))' failed.
```

tungld commented on test/modellib/ConvModel.cpp:
```diff
@@ -99,7 +99,7 @@ bool Conv2DLibBuilder::build() {
 
```
Perhaps, you shoudl change `-1` in Line 63-64 to `ShapedType::kDynamic`too.

Lines 63-64:
```
  if (isDynamic)
    N1 = CIn1 = COut1 = H1 = W1 = -1;
```

christopherbate replied:
Thanks. I did do this yesterday in Conv and the other model builder files, but I'm still working on another issue.

christopherbate commented:
I was able to fix the immediate issue and get numerical tests running without crashing, but there's another bug that's causing the numerical results to fail. I believe this is currently the cause: 

```mlir
// -----// IR Dump After onnx_mlir::(anonymous namespace)::ShapeInferencePass (shape-inference) //----- //
module attributes {llvm.data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", llvm.target_triple = "x86_64-unknown-linux-gnu"} {
  func.func @main_graph(%arg0: tensor<?x?x?x?xf32>, %arg1: tensor<1x1x1x1xf32>) -> tensor<?x1x?x?xf32> {
    %0 = "onnx.NoValue"() {value} : () -> none
    %1 = "onnx.Conv"(%arg0, %arg1, %0) {auto_pad = "VALID", dilations = [1, 1], group = 1 : si64, kernel_shape = [1, 1], pads = [0, 0, 0, 0], strides = [1, 1]} : (tensor<?x?x?x?xf32>, tensor<1x1x1x1xf32>, none) -> tensor<?x1x?x?xf32>
    return %1 : tensor<?x1x?x?xf32>
  }
  "onnx.EntryPoint"() {func = @main_graph} : () -> ()
}


// -----// IR Dump After (anonymous namespace)::ConvOptONNXToONNXPass (conv-opt-onnx) //----- //
func.func @main_graph(%arg0: tensor<?x?x?x?xf32>, %arg1: tensor<1x1x1x1xf32>) -> tensor<?x1x?x?xf32> {
  %0 = "onnx.NoValue"() {value} : () -> none
  %1 = "onnx.Shape"(%arg0) {start = 0 : si64} : (tensor<?x?x?x?xf32>) -> tensor<4xi64>
  %2 = onnx.Constant dense<-1> : tensor<1xi64>
  %3 = onnx.Constant dense<0> : tensor<1xi64>
  %4 = onnx.Constant dense<0> : tensor<1xi64>
  %5 = onnx.Constant dense<2> : tensor<1xi64>
  %6 = onnx.Constant dense<1> : tensor<1xi64>
  %7 = "onnx.Slice"(%1, %4, %5, %3, %6) : (tensor<4xi64>, tensor<1xi64>, tensor<1xi64>, tensor<1xi64>, tensor<1xi64>) -> tensor<2xi64>
  %8 = "onnx.Concat"(%7, %2) {axis = 0 : si64} : (tensor<2xi64>, tensor<1xi64>) -> tensor<3xi64>
  %9 = "onnx.Reshape"(%arg0, %8) {allowzero = 0 : si64} : (tensor<?x?x?x?xf32>, tensor<3xi64>) -> tensor<?x?x?xf32>
  %10 = "onnx.Shape"(%arg1) {start = 0 : si64} : (tensor<1x1x1x1xf32>) -> tensor<4xi64>
  %11 = onnx.Constant dense<-1> : tensor<1xi64>
  %12 = onnx.Constant dense<0> : tensor<1xi64>
  %13 = onnx.Constant dense<0> : tensor<1xi64>
  %14 = onnx.Constant dense<1> : tensor<1xi64>
  %15 = onnx.Constant dense<1> : tensor<1xi64>
  %16 = "onnx.Slice"(%10, %13, %14, %12, %15) : (tensor<4xi64>, tensor<1xi64>, tensor<1xi64>, tensor<1xi64>, tensor<1xi64>) -> tensor<1xi64>
  %17 = "onnx.Concat"(%16, %11) {axis = 0 : si64} : (tensor<1xi64>, tensor<1xi64>) -> tensor<2xi64>
  %18 = "onnx.Reshape"(%arg1, %17) {allowzero = 0 : si64} : (tensor<1x1x1x1xf32>, tensor<2xi64>) -> tensor<1x?xf32>
  %19 = "onnx.MatMul"(%18, %9) : (tensor<1x?xf32>, tensor<?x?x?xf32>) -> tensor<0x1x?xf32>
  %20 = "onnx.Shape"(%arg0) {start = 0 : si64} : (tensor<?x?x?x?xf32>) -> tensor<4xi64>
  %21 = "onnx.Shape"(%arg1) {start = 0 : si64} : (tensor<1x1x1x1xf32>) -> tensor<4xi64>
  %22 = onnx.Constant dense<0> : tensor<1xi64>
  %23 = onnx.Constant dense<0> : tensor<1xi64>
  %24 = onnx.Constant dense<1> : tensor<1xi64>
  %25 = onnx.Constant dense<1> : tensor<1xi64>
  %26 = "onnx.Slice"(%20, %23, %24, %22, %25) : (tensor<4xi64>, tensor<1xi64>, tensor<1xi64>, tensor<1xi64>, tensor<1xi64>) -> tensor<1xi64>
  %27 = onnx.Constant dense<0> : tensor<1xi64>
  %28 = onnx.Constant dense<0> : tensor<1xi64>
  %29 = onnx.Constant dense<1> : tensor<1xi64>
  %30 = onnx.Constant dense<1> : tensor<1xi64>
  %31 = "onnx.Slice"(%21, %28, %29, %27, %30) : (tensor<4xi64>, tensor<1xi64>, tensor<1xi64>, tensor<1xi64>, tensor<1xi64>) -> tensor<1xi64>
  %32 = onnx.Constant dense<0> : tensor<1xi64>
  %33 = onnx.Constant dense<2> : tensor<1xi64>
  %34 = onnx.Constant dense<4> : tensor<1xi64>
  %35 = onnx.Constant dense<1> : tensor<1xi64>
  %36 = "onnx.Slice"(%20, %33, %34, %32, %35) : (tensor<4xi64>, tensor<1xi64>, tensor<1xi64>, tensor<1xi64>, tensor<1xi64>) -> tensor<2xi64>
  %37 = "onnx.Concat"(%26, %31, %36) {axis = 0 : si64} : (tensor<1xi64>, tensor<1xi64>, tensor<2xi64>) -> tensor<4xi64>
  %38 = "onnx.Reshape"(%19, %37) {allowzero = 0 : si64} : (tensor<0x1x?xf32>, tensor<4xi64>) -> tensor<?x1x?x?xf32>
  return %38 : tensor<?x1x?x?xf32>
}
```

(matmul result has a 0 in the first dim)

Note that I also had to correct use of `int` in several locations: if you do `int a = static_cast<int>( someShapedType.getDimSize(2) )`, and it happens that `someShapedType.getDimSize(2) == ShapedType::kDynamic` , then `a` will be `0`.

I wasn't able to find the root cause yesterday after an hour of experimenting. I'll give it another shot today, but any help would be appreciated.

christopherbate commented:
Recommend adding `Werror=constant-conversion` to catch issues like this:

```
test/modellib/RNNModel.cpp:45:10: warning: implicit conversion from 'const int64_t' (aka 'const long') to 'int' changes value from -9223372036854775808 to 0 [-Wconstant-conversion]
    S1 = ShapedType::kDynamic;
```

NathanielMcVicar commented:
@jenkins-droid test this please

NathanielMcVicar commented:
Apparently, I don't have permission to start a test. Hopefully someone that does sees this soon!

philass commented:
@jenkins-droid test this please

christopherbate commented:
I'm passing all `check-onnx-mlir`, `check-onnx-backend`, `check-onnx-backend-numerical`, `check-onnx-backend-dynamic` locally. Looks like CI failing due to auth issue?

Once someone gives me the OK and it passes CI I'll clang-format all changed files.

AlexandreEichenberger commented:
@jenkins-droid test this please

AlexandreEichenberger commented:
From the Windows CI, it looks like there is a `make check-docs` error. This runs tests that ensure the doc (extract of code in the doc) is in sync with the scripts and code. Run the command and hopefully you will be able to identify which script is impacted, and then git diff with the old version should help you find out how to update the doc pages. Not sure what could have triggered it.

AlexandreEichenberger commented:
merging latest change, which also restarts the CIs.

sstamenova commented:
@christopherbate Once this update is complete, if you found anything particularly useful or not that you think would be good to pass along, please add it to the update instructions.

Thank you for doing the update!

AlexandreEichenberger commented:
On the Jenkins, we build both the debug and the release version. Release is a bit more picky about included libraries. You should be able to reproduce it by doing so... here is my personal script
```bash
rm -rf onnx-mlir/build
mkdir onnx-mlir/build && cd onnx-mlir/build
if [[ -z "$pythonLocation" ]]; then
  cmake -G "Unix Makefiles" -DCMAKE_BUILD_TYPE=Release -DCMAKE_CXX_COMPILER=/usr/bin/c++ -DMLIR_DIR=/workdir/llvm-project/build/lib/cmake/mlir -DONNX_MLIR_ACCELERATORS=NNPA ..
else
  cmake -G "Unix Makefiles" -DCMAKE_BUILD_TYPE=Release -DCMAKE_CXX_COMPILER=/usr/bin/c++ -DMLIR_DIR=/workdir/llvm-project/build/lib/cmake/mlir -DONNX_MLIR_ACCELERATORS=NNPA -DPython3_ROOT_DIR=$pythonLocation ..
fi
cmake --build .

# Run lit tests:
export LIT_OPTS=-v
cmake --build . --target check-onnx-lit
```

christopherbate commented:
Looks like it's due to building LLVM with BUILD_SHARED_LIBS=ON vs OFF. I didn't think about testing that. I'll rebuild and let you know.

AlexandreEichenberger commented:
@jenkins-droid test this please

AlexandreEichenberger commented:
@jenkins-droid test this please

AlexandreEichenberger commented:
maybe still some lib issues?
```
+ /var/lib/jenkins/workspace/ONNX-MLIR-Pipeline-Docker-Build@pr_1977/.buildbot/jenkins-verify-user-image.py
[2023-01-31 11:04:34][069] INFO: docker run --rm -u 111:117 -v /var/lib/jenkins/workspace/ONNX-MLIR-Pipeline-Docker-Build@pr_1977/mnist-8:/workdir/mnist-8 onnxmlirczar/onnx-mlir:1977 --EmitLib /workdir/mnist-8/mnist-8.onnx
onnx-mlir: error while loading shared libraries: libVhloOps.so.16git: cannot open shared object file: No such file or directory
[Pipeline] }
```

tungld commented:
@jenkins-droid test this please

AlexandreEichenberger commented:
Looks like you did it. Please update the format and we will try to hold on pushing code until you do

AlexandreEichenberger commented:
@christopherbate Thanks for this massive undertaking. Please pat yourself on the back on behalf of all of us.

sstamenova commented:
> @sstamenova Thanks for offering to diff this PR with your own local copy where you also removed the `-1` for dynamic variables. It will provide a second validation that all is in order, thanks.

I took a brief look and found a few differences with changes needed in both repos. I'll send a PR as I get a chance.
